packages/jupyter-ai/jupyter_ai/document_loaders/directory.py

```diff
@@ -1,54 +1,73 @@
-import logging
-from typing import Dict, List, Optional, Union
+from typing import List, Optional
+from pathlib import Path
+import hashlib
+import itertools
+
+import ray
 
 from langchain.document_loaders.base import BaseLoader
-from langchain.document_loaders.directory import FILE_LOADER_TYPE, _is_visible
-from langchain.document_loaders.unstructured import UnstructuredFileLoader
 from langchain.schema import Document
-from wcmatch.pathlib import Path
-
-
-logger = logging.getLogger(__name__)
+from langchain.document_loaders.directory import _is_visible
+from langchain.text_splitter import (
+    RecursiveCharacterTextSplitter, TextSplitter,
+)
 
+@ray.remote
+def path_to_doc(path):
+    with open(str(path), 'r') as f:
+        text = f.read()
+        m = hashlib.sha256()
+        m.update(text.encode('utf-8'))
+        metadata = {'path': str(path), 'sha256': m.digest(), 'extension': path.suffix}
+        return Document(page_content=text, metadata=metadata)
 
-class DirectoryLoader(BaseLoader):
-    """Loading logic for loading documents from a directory."""
+class ExcludePattern(Exception):
+    pass
+    
+def iter_paths(path, extensions, exclude):
+    for p in Path(path).rglob('*'):
+        if p.is_dir():
+            continue
+        if not _is_visible(p.relative_to(path)):
+            continue
+        try:
+            for pattern in exclude:
+                if pattern in str(p):
+                    raise ExcludePattern()
+        except ExcludePattern:
+            continue
+        if p.suffix in extensions:
+            yield p
 
+class RayRecursiveDirectoryLoader(BaseLoader):
+    
     def __init__(
         self,
-        path: str,
-        glob: Union[str, List[str]] = "**/[!.]*",
-        silent_errors: bool = False,
-        load_hidden: bool = False,
-        loader_cls: FILE_LOADER_TYPE = UnstructuredFileLoader,
-        loader_kwargs: Optional[Dict] = None,
-        recursive: bool = False,
+        path,
+        extensions={'.py', '.md', '.R', '.Rmd', '.jl', '.sh', '.ipynb', '.js', '.ts'},
+        exclude={'.ipynb_checkpoints', 'node_modules', 'lib', 'build'}
```

Comment:
Might be good to ignore `.git` and `.DS_Store` as well

Reply:
Could we parse `.gitignore` and use that for our denylist by default? This can be addressed in a follow-up PR, since the problem of "over-indexing" is generally not immediately obvious to end users.

Reply:
Yep, let's explore this in a follow up PR.